File: app/babycare/camera_local_file.py
import cv2
import datetime
import time

# 获取摄像头
cap = cv2.VideoCapture(0)

if not cap.isOpened():
    raise Exception("无法打开摄像头")

# 设置帧率和分辨率
cap.set(cv2.CAP_PROP_FPS, 5)
cap.set(cv2.CAP_PROP_FRAME_WIDTH, 1280)
cap.set(cv2.CAP_PROP_FRAME_HEIGHT, 720)


def get_vidoe_out():
    now = datetime.datetime.now(datetime.timezone(datetime.timedelta(hours=8)))
    filename = now.strftime("%Y-%m-%d_%H-%M-%S.avi")
    out = cv2.VideoWriter(
        filename,
        cv2.VideoWriter_fourcc(*"XVID"),
        5,
        (1280, 720),
    )
    return out


try:
    split_interval_s = 10.0  # 设定每个录像文件的时长为10秒

    out = get_vidoe_out()
    start_time = time.time()
    while True:
        ret, frame = cap.read()

        if not ret:
            raise Exception("无法读取视频帧数据")

        elapsed_time = time.time() - start_time  # 计算已经过去的录像时间

        if elapsed_time > split_interval_s:  # 如超过设定的interval，则创建新的录像文件
            if out is not None:
                out.release()
            else:
                raise Exception("OpenCV VideoWriter对象为空，无法保存视频数据")

            out = get_vidoe_out()

            start_time = time.time()  # 更新start_time

        out.write(frame)

finally:
    cap.release()
    if out is not None:
        out.release()

File: app/babycare/camera_local_file.cpp
#include <opencv2/opencv.hpp>
#include <string>

int main() {
    cv::VideoCapture cap(0); // open the video camera no. 0

    if (!cap.isOpened()) { // 如果没有相机
        std::cout << "Cannot open the video cam" << std::endl;
        return -1;
    }

    double dWidth = cap.get(cv::640);       // get capture width
    double dHeight = cap.get(cv::480);     // get capture height
    std::cout << "Frame size : " << dWidth << " x " << dHeight << std::endl;
  
    int frame_rate = 30;   // set your desired frame rate
    auto codec = cv::VideoWriter::fourcc(*"XVID");

    time_t start = time(0);
    long vid_duration_sec = 10;
    int idx = 0;

    while (true) {
        cv::Mat frame;

        bool bSuccess = cap.read(frame);   // read a new frame from video

        if (!bSuccess) {                   //if not successful, break loop
            std::cout << "Cannot read a frame from video stream" << std::endl;
            break;
        }

         /*we’re constantly grabbing frames and checking for keypresses afterwards*/
        if (difftime(time(0), start) >= vid_duration_sec) {
            idx++;
            start = time(0);
        }

        std::string filename = "Myvideo_" + std::to_string(idx) + ".avi";
        cv::VideoWriter record(filename,
                               codec,
                               frame_rate,
                               cv::Size(dWidth, dHeight));

        if (!record.isOpened()) {
            std::cerr << "Could not write video file" << std::endl;
            return -1;
        }
        
        record.write(frame);

        char c = (char)cv::waitKey(33);
        if (c == 27) {              // Press 'esc' to stop capturing.
            break;
        }
    }

    cap.release();
    return 0;
}

// sudo apt install libopencv-dev
// g++ yourfile.cpp -o outputfile `pkg-config --cflags --libs opencv4`